<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jason Yeung" />

<meta name="date" content="2021-12-25" />

<title>Analyzing ViReMa Output in R</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Visualizing Viral Recombination: Analyzing ViReMa Output</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="rtutorial.html">Tutorial with R</a>
</li>
<li>
  <a href="shinytutorial.html">Tutorial with Shiny</a>
</li>
<li>
  <a href="https://routhlab.shinyapps.io/ViReMaShiny/">ViReMa Web App</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="mailto:&lt;jasonyeung012@gmail.com&gt;">
    <span class="fa fa-envelope fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/routhlab/ViReMaShiny">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Analyzing ViReMa Output in R</h1>
<h4 class="author">Jason Yeung</h4>
<h4 class="date">12/25/2021</h4>

</div>


<hr />
<div id="getting-started-with-virema-in-r" class="section level2">
<h2><strong>Getting Started with ViReMa in R</strong></h2>
<p>This tutorial will guide users through analyzing viral recombination data from <a href="https://bioconda.github.io/recipes/virema/README.html">ViReMa</a> in R. While the Shiny app linked in the navigation bar provides quick analysis without knowing how to code, learning how to analyze ViReMa data with R gives you full control of plotting and analysis! If you are completely new to R, download the latest version of RStudio at <a href="http://rstudio.com" class="uri">http://rstudio.com</a> before following along.</p>
<p><br></p>
</div>
<div id="setting-up" class="section level2">
<h2><strong>Setting Up</strong></h2>
<p>The first step will be to install packages:</p>
<pre class="r"><code>install.packages(c(&quot;ggplot2&quot;,&quot;tidyverse&quot;,&quot;circlize&quot;))</code></pre>
<p>This allows us to use popular functions we’ll need for data visualization.</p>
<p><br></p>
<p>Now, let’s <strong>load some example data</strong> I’ve provided:</p>
<pre class="r"><code>#Loading libraries
library(ggplot2)
library(tidyverse)

#We&#39;ve provided some sample data for the purposes of this tutorial
#Load provided example data

txt1 = &quot;https://raw.githubusercontent.com/jayeung12/jayeung12.github.io/main/files/SARS-CoV-2_1.txt&quot;
txt2 = &quot;https://raw.githubusercontent.com/jayeung12/jayeung12.github.io/main/files/SARS-CoV-2_2.txt&quot;  
txtdf1 = read.table(txt1,skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)
txtdf2 = read.table(txt2,skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)

data = rbind(txtdf1,txtdf2)</code></pre>
<p><br> <strong>Alternatively, load in your own data</strong> and follow along with the rest of the tutorial. Naturally, the plots will not be exactly the same as the ones we show.</p>
<pre class="r"><code>#Or you can use your own data
#Load own data from local directory

#This loads in a single file
#skip excludes the first line of the file
data = read.table(&quot;--replace with the file path--.bed&quot;,skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)

#To load in multiple files all in the same directory, we can use the following:

filenames = list.files(&quot;--replace with the directory--&quot;, pattern=&quot;*.bed&quot;, full.names=TRUE)
filelist = list()
for(i in 1:length(filenames)){
  filelist[[i]] =   read.table(filenames[i],skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)
}

#Combines all of the files, row by row
data = do.call(rbind, filelist)</code></pre>
<p><br> The <strong>data is in .BED6 format but the file extension may just say .BED or .TXT</strong>. If that’s the case, you will need to match the file extension in the above code to load in your own data.</p>
<p><strong>Your files need to have at least 6 columns</strong> - these correspond to the reference sequence, the donor site (start), the acceptor site (end), a description field, reads, and strandedness. If you selected other options during ViReMa analysis, you may have extra fields like reads at the acceptor site/donor site or nucleotide sequences at those sites.</p>
<p><img src="files/bedColumns.PNG" alt="Example format for file columns" /> <br><br></p>
<p>This figure below illustrates what each column corresponds to: <br><br></p>
<p><img src="files/ViremaShinyFigure.png" alt="Figure for file columns and sequence" /> <br><br></p>
<p><br></p>
</div>
<div id="plotting-recombination-events" class="section level2">
<h2><strong>Plotting Recombination Events</strong></h2>
<div id="scatterplot" class="section level3">
<h3><strong>Scatterplot</strong></h3>
<p>Now that we have the data loaded in, <strong>let’s make our first plot</strong>. We will use ‘ggplot2’ to make a scatterplot with acceptor and donor sites on the axes. With multiple .BED files for different viral samples, identical recombination events may happen across multiple samples. We want to visualize how often events occur across multiple samples.</p>
<pre class="r"><code>#Before plotting, we need to make a few more variables.
#V11 is the number of samples an event appears in
#V5 is updated to reflect the sum of the reads across samples
#Make sure to ungroup at the end - it may cause issues down the line
dataCombo = data%&gt;% group_by(V2,V3,V4) %&gt;% mutate(V11=n()) %&gt;% mutate(V5=sum(V5 %&gt;% as.numeric())) %&gt;% ungroup()

#Making the plot
#The size of the points are controlled by the number of reads - &#39;size = V5&#39;
#The color is controlled by the number of isolates (aka .BED files) each event appears in - &#39;color = V11&#39;
#Change the colors at &#39;scale_color_gradient&#39;
p = ggplot(dataCombo,aes(x = V3, y = V2, color = V11)) +
            geom_point(aes(size = V5),alpha = 0.6) +
            scale_size_continuous(&quot;Read Count&quot;,range = c(0.4, 3))+
            coord_cartesian() +
          ylab(&quot;Donor Site&quot;) +
          xlab(&quot;Acceptor Site&quot;) +
            theme_classic(base_size = 18) +
            scale_color_gradient(&quot;# of Isolates&quot;,low = &quot;#87f6ff&quot;, high = &quot;#f786ff&quot;)
p</code></pre>
<p><img src="files/scatter-1.png" alt="A scatterplot" /></p>
<p><br></p>
</div>
<div id="coverage-plot" class="section level3">
<h3><strong>Coverage Plot</strong></h3>
<p><strong>This plot serves as a coverage plot.</strong> It takes each recombination event and adds the number of reads at nucleotide indexes between the donor and acceptor site. Depending on if deletion or duplication events are being plotted, this shows areas of the genome which are either frequently removed or included in recombination events. <br></p>
<pre class="r"><code>#Making a coverage plot for only &#39;deletion&#39; events - &#39;filter(V2&lt;V3)&#39; meaning the Acceptor is after the Donor site
#To plot duplications, simply switch V2 and V3 - so &#39;filter(V3&lt;V2)&#39;

Del = dataCombo%&gt;% select(V2,V3,V4,V5) %&gt;% filter(V2 &lt; V3) %&gt;% mutate(id = paste0(V2,V3,V4,collapse = &quot;,&quot;)) %&gt;% group_by(id) %&gt;% filter(row_number() == 1) %&gt;% ungroup() %&gt;% rowwise() %&gt;% mutate(seq = list(seq(from = V2,to = V3,by = 1))) %&gt;% unnest(cols = seq) %&gt;% group_by(seq) %&gt;% mutate(n=sum(V5) )

#This may take awhile depending on the binwidth you choose
#The lower the binwidth the greater nucleotide resolution for the histogram
p = ggplot(Del,aes(x = seq)) + geom_histogram(binwidth = 100, fill = &quot;lightblue&quot;) +
        ylab(&quot;Reads&quot;) +
        xlab(&quot;Base Position&quot;) +
        theme_classic(base_size = 18)
p</code></pre>
<p><img src="files/coverage-1.png" alt="A coverage plot" /> <br><br></p>
</div>
<div id="circos-plot" class="section level3">
<h3><strong>Circos Plot</strong></h3>
<p><strong>This plot is a Circos plot made using the ‘circlize’ package.</strong> It allows for recombination events to be visualized alongside annotations or other quantitative data. To fully understand how plotting a Circos plot with ‘circlize’ works, read this documentation: <a href="https://jokergoo.github.io/circlize_book/book/" class="uri">https://jokergoo.github.io/circlize_book/book/</a>. There are many types of data that could potentially be plotted alongside recombination events, some of which we won’t touch on in this tutorial. <br></p>
<p><strong>There are multiple places to find annotations for your viral genome.</strong> NCBI has annotations associated with sequences. Below shows this process for a popular SARS-CoV-2 reference sequence. <br><br> <img src="files/exportNCBIanno.gif" alt="Accessing the annotations at UCSC" /></p>
<p><br> The UCSC Genome browser website has a comprehensive collection of annotations for SARS-CoV-2. Access the “Data Integrator” at <a href="https://genome.ucsc.edu/cgi-bin/hgIntegrator" class="uri">https://genome.ucsc.edu/cgi-bin/hgIntegrator</a>. Below shows the process of exporting data:</p>
<p><img src="files/exportUCSCanno.gif" alt="Accessing the annotations at UCSC" /> <br><br> Then, copy the data into Notepad or another text editor and save it as a .BED or .TXT file. <br><br> <strong>The annotations also need certain columns to be plottable:</strong> <br><br> <img src="files/annoColumns.png" alt="Example format for annotations" /> <br><br> They correspond to the ref seq, start, end, name, value, and strand. Only the first 4 are necessary for our purposes. <strong>Important note - make sure the reference sequence name matches a reference sequence name in your .BED files. The plot will not work otherwise!</strong> Let’s load in some annotations.</p>
<pre class="r"><code>#This plot requires the circlize package
#Gu, Z. (2014) circlize implements and enhances circular visualization in R. Bioinformatics. DOI: 10.1093/bioinformatics/btu393

library(circlize)

url = &quot;https://raw.githubusercontent.com/jayeung12/jayeung12.github.io/main/files/sars-cov-2Anno.BED&quot;  
annotations = read.table(url, sep = &quot;\t&quot;,stringsAsFactors = FALSE, quote = &quot;&quot;)


#Run this if you want to import your own annotations instead
annotations = read.table(&quot;--replace with the file path--.bed&quot;, sep = &quot;\t&quot;,stringsAsFactors = FALSE, quote = &quot;&quot;)</code></pre>
<p>If we were to plot the annotations on the same row and track, there is the possibility they will overlap. We’ll need to run a bit of code to properly assign rows to the annotations.</p>
<pre class="r"><code>#Making a new variable called row
annotations$row = NA
annotations$row[1] = 1

annotationsOrd = annotations %&gt;% select(V1,V2,V3,V4,row) %&gt;% arrange(V2)
for(i in 2:nrow(annotationsOrd)){
  #The 400 in the next line provides a buffer space between genes - play around with it for aesthetics
  noOverlaps = annotationsOrd %&gt;% na.omit %&gt;% group_by(row) %&gt;% mutate(Overlap=any( V3+400 &gt;= annotationsOrd$V2[i] ) ) %&gt;% filter(Overlap==F) %&gt;% na.omit%&gt;% arrange(row,desc())
  
  if(nrow(noOverlaps!=0)){
    annotationsOrd$row[i] = noOverlaps$row[1]
  }else{
    annotationsOrd$row[i] = max(annotationsOrd$row,na.rm = T)+1
  }
    
}
annotationsOrd$row[1] = 1

#Let&#39;s look at the results
annotationsOrd
annotationsOrd$row</code></pre>
<p><br><br> On to plotting the recombination events! We are going to use directed arrows to depict recombination events.</p>
<pre class="r"><code>#Splitting the events 
#If you want to plot a subset of events (say only the deletions), make a new dataframe and replace dataCombo everywhere below
b1 = dataCombo %&gt;% select(V1,V2)
b2 = dataCombo %&gt;% select(V1,V3)

#A scaling factor based on the number of rows from annotationsOrd
#This sets the height of the boxes for the annotations
scaling = 100/max(annotationsOrd$row,na.rm = T)

#The actual plotting!
circos.par(&quot;track.height&quot; = 0.2)
#xlim is technically optional - however, it will guess the limits based on the start and end of events you include if you omit it
circos.initialize(sectors =  dataCombo$V1, dataCombo$V3, xlim = c(0,30000))
circos.track(ylim = c(1,120),
             panel.fun = function(x, y) {
               circos.axis(labels.cex = 0.6)
               circos.genomicLink(b1,b2,col = add_transparency(&quot;black&quot;, 0.95), arr.type = &quot;triangle&quot;,directional = 1,
               border = NA)
             })

for(i in 1:length(unique(annotations$V1)) )
{
  seqs=annotationsOrd %&gt;% filter(V1==unique(annotationsOrd$V1)[i])
  circos.rect(sector.index = seqs$V1[1], seqs$V2, (seqs$row-1)*scaling, seqs$V3, seqs$row*scaling,col=&quot;#FF0000&quot;)
  circos.text(sector.index = seqs$V1[1], x=(seqs$V2+seqs$V3)/2, y=((seqs$row-1)*scaling)+20, seqs$V4, cex=.6, facing = &quot;bending.inside&quot;)
}</code></pre>
<p><img src="files/circos-1.PNG" alt="A Circos plot" /> <br><br></p>
<p><strong>What if we want to color types of recombination events differently?</strong> It’s a matter of filtering our data first and then adding the arrows in sequence.</p>
<pre class="r"><code>b1Dup = dataCombo %&gt;% filter(V2&gt;V3) %&gt;%  select(V1,V2)
b2Dup = dataCombo %&gt;% filter(V2&gt;V3) %&gt;% select(V1,V3)

b1Del = dataCombo %&gt;% filter(V2&lt;V3) %&gt;%  select(V1,V2)
b2Del = dataCombo %&gt;% filter(V2&lt;V3) %&gt;% select(V1,V3)

circos.par(&quot;track.height&quot; = 0.2)
circos.initialize(sectors =  dataCombo$V1,dataCombo$V3,xlim = c(0,30000))
circos.track(ylim = c(1,120),
             panel.fun = function(x, y) {
               for(i in 1:length(unique(genes$V1)) )
               {
                 circos.axis(labels.cex = 0.6)
                 seqs=genesOrd %&gt;% filter(V1==unique(genesOrd$V1)[i])
                 circos.rect(sector.index = seqs$V1[1], seqs$V2, (seqs$row-1)*scaling, seqs$V3, seqs$row*scaling,col=&quot;#FF0000&quot;)
                 circos.text(sector.index = seqs$V1[1], x=(seqs$V2+seqs$V3)/2, y=((seqs$row-1)*scaling)+20, seqs$V4, cex=.6, facing = &quot;bending.inside&quot;)
                 
               }
             })
#Just add them one after another with different colors
circos.genomicLink(b1Dup,b2Dup,col = add_transparency(&quot;black&quot;, 0.95), arr.type = &quot;triangle&quot;,directional = 1,
                   border = NA)
circos.genomicLink(b1Del,b2Del,col = add_transparency(&quot;#F2AD00&quot;, 0.95), arr.type = &quot;triangle&quot;,directional = 1,
                   border = NA)</code></pre>
<p><img src="files/circos-3.PNG" alt="A Circos plot with different arrow colors" /> <br><br></p>
<p><strong>Let’s explore adding more tracks.</strong> Here, we add simulated quantitative data as a histogram.</p>
<pre class="r"><code>#The same plot but with a new track
circos.par(&quot;track.height&quot; = 0.2)
circos.initialize(sectors =  dataCombo$V1,dataCombo$V3,xlim = c(0,30000))
circos.track(ylim = c(1,120),
             panel.fun = function(x, y) {
               for(i in 1:length(unique(annotations$V1)) )
               {
                 circos.axis(labels.cex = 0.6)
                 seqs=annotationsOrd %&gt;% filter(V1==unique(annotationsOrd$V1)[i])
                 circos.rect(sector.index = seqs$V1[1], seqs$V2, (seqs$row-1)*scaling, seqs$V3, seqs$row*scaling,col=&quot;#FF0000&quot;)
                 circos.text(sector.index = seqs$V1[1], x=(seqs$V2+seqs$V3)/2, y=((seqs$row-1)*scaling)+20, seqs$V4, cex=.6, facing = &quot;bending.inside&quot;)
                 
               }
             })
#Here&#39;s the new track - it&#39;s just simulated data
circos.track(ylim = c(-1, 1), panel.fun = function(x, y) {
  xvalue = 1:30000
  xvalue=xvalue[seq(1, 30000, 20)]
  yvalue=runif(1500,-1,1)
  circos.lines(type=&#39;s&#39;, area = T, x=xvalue, y=yvalue, col=&quot;lightblue&quot;)
})
circos.genomicLink(b1,b2,col = add_transparency(&quot;black&quot;, 0.95), arr.type = &quot;triangle&quot;,directional = 1,
                   border = NA)</code></pre>
<p><img src="files/circos-2.PNG" alt="A Circos plot with 2 tracks" /> <br><br> <strong>What if you want to plot a subsection of the total plot?</strong> We have to do two things: 1. Limit the arrows so that both the starts and ends occur within the limits we set and 2. Limit the annotations so only annotations within the limits show.</p>
<pre class="r"><code>b1bool=vector()
b2bool=vector()
b1$id = 1:nrow(b1)
b2$id = 1:nrow(b2)

#Let&#39;s set our xlimits (nucleotide coordinates)
xLims=c(10000,25000)

#Does the start occur between the limits?
hold = b1 %&gt;% dplyr::filter( between(V2,xLims[1],xLims[2]) )
b1bool = c(b1bool,hold$id)

#Does the end?
hold = b2 %&gt;% dplyr::filter( between(V3,xLims[1],xLims[2]) )
b2bool = c(b2bool,hold$id)


#We update b1 to include the events where both the start and end occur within the limits
b1=b1[intersect(b1bool,b2bool),1:2]
b2=b2[intersect(b1bool,b2bool),1:2]


#We do a similar thing with the annotations
finalAnno=data.frame()

#identifies annotations with starts or ends in between the xlimits
#changes the ends to fit in between the xlimits
hold = annotationsOrd  %&gt;% dplyr::filter( between(V2,xLims[1],xLims[2])|between(V3,xLims[1],xLims[2]) ) %&gt;% mutate(V2=ifelse(V2&lt;xLims[1],xLims[1],V2)) %&gt;% mutate(V3=ifelse(V3&gt;xLims[2],xLims[2],V3))
finalAnno=rbind(finalAnno,hold)
#identifies annotations with starts and ends more extreme than xlimits
#changes the ends to fit in between the xlimits
hold = annotationsOrd  %&gt;% dplyr::filter(V2&lt;xLims[1]&amp;V3&gt;xLims[2]) %&gt;% mutate(V2 = xLims[1]) %&gt;% mutate(V3 = xLims[2])
finalAnno=rbind(finalAnno,hold)



#And finally, re-running the plot
scaling = 100/max(finalAnno$row,na.rm = T)

circos.par(&quot;track.height&quot; = 0.2)
#xlim is set to xLims
circos.initialize(sectors =  dataCombo$V1, dataCombo$V3, xlim = xLims)
circos.track(ylim = c(1,120),
             panel.fun = function(x, y) {
               circos.axis(labels.cex = 0.6)
               circos.genomicLink(b1,b2,col = add_transparency(&quot;black&quot;, 0.95), arr.type = &quot;triangle&quot;,directional = 1, border = NA)
             })

for(i in 1:length(unique(finalAnno$V1)) ){
  seqs=finalAnno %&gt;% filter(V1==unique(finalAnno$V1)[i])
  circos.rect(sector.index = seqs$V1[1], seqs$V2, (seqs$row-1)*scaling, seqs$V3, seqs$row*scaling,col=&quot;#FF0000&quot;)
  circos.text(sector.index = seqs$V1[1], x=(seqs$V2+seqs$V3)/2, y=((seqs$row-1)*scaling)+20, seqs$V4, cex=.6, facing = &quot;bending.inside&quot;)
}</code></pre>
<p><img src="files/circos-4.PNG" alt="A Circos plot with a subsection" /> <br><br> There are much crazier features you can add to the Circos plot using ‘circlize’. Again, check out <a href="https://jokergoo.github.io/circlize_book/book/" class="uri">https://jokergoo.github.io/circlize_book/book/</a> to get ideas for what’s possible.</p>
<p><br><br></p>
</div>
<div id="nucleotide-usage-plot" class="section level3">
<h3><strong>Nucleotide Usage Plot</strong></h3>
<p><strong>This plot shows the frequency of nucleotides proximal and distal to recombination sites.</strong> In order to generate these plots, you had to select the option to produce this output when performing the analysis with ViReMA. The necessary columns look like this: <br><br> <img src="files/nucleotideColumns.PNG" alt="Example format for nucleotide columns" /> <br><br></p>
<p>Let’s make a plot looking at nucleotides around the donor site.</p>
<pre class="r"><code>#The nucleotide at position 0 will be the last nucleotide before the recombination event - before the &#39;|&#39; symbol
donor = dataCombo %&gt;% select(V2,V3,V4,V5,V6,V7,V9)%&gt;% mutate(letters = str_split(V9, &quot;&quot;)) %&gt;%
        unnest(cols = letters) %&gt;%filter(letters!=&quot;|&quot;) %&gt;% 
        group_by(V2,V3,V4,V5,V6,V7, V9) %&gt;%
        mutate(position = row_number()-25) %&gt;% ungroup()%&gt;% count(letters,position) 
      
p = ggplot(donor,aes(x=position, y=n, group=letters)) +
        geom_line(aes(color=letters),size=1)+
        geom_vline(xintercept = 0, linetype=&quot;dashed&quot;, 
                     color = &quot;gray&quot;, size=1)+
        ylab(&quot;# of Sequences&quot;) +
        xlab(&quot;Base Position Relative to Donor&quot;) +
        theme_classic(base_size = 18)
p</code></pre>
<p><img src="files/donor-1.png" alt="A donor nucleotide plot" /> <br> We can see a clear nucleotide bias for ‘T’ at the donor site of unique recombination events. What if we wanted to account for the number of reads for each recombination event? You could divide the read number for each event by the mean of the read numbers at the donor and acceptor sites.</p>
<p><br><br></p>
</div>
</div>
<div id="segmented-viruses" class="section level2">
<h2><strong>Segmented Viruses</strong></h2>
<p><strong>Plotting events in segmented viruses requires some changes to the code we used above.</strong> .BED files will contain multiple reference sequence IDs in the first column corresponding to the different segments of the viral genome. Note that .BED files can only depict events that start and end on the same reference sequence. .BEDPE files which can also be generated by the ViReMa script are the file format for recombination events between viral segments or viral-host nucleic acids.</p>
<p>We can also make the Circos plot with multiple reference sequences.</p>
<p><br><br></p>
</div>
<div id="session-info" class="section level2">
<h2><strong>Session Info</strong></h2>
<pre class="r"><code>sessionInfo()

# R version 4.0.4 (2021-02-15)
# Platform: x86_64-w64-mingw32/x64 (64-bit)
# Running under: Windows 10 x64 (build 19042)
# 
# Matrix products: default
# 
# locale:
# [1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252
# [3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C
# [5] LC_TIME=English_United States.1252
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base
# 
# other attached packages:
#  [1] circlize_0.4.13 forcats_0.5.1   stringr_1.4.0   dplyr_1.0.7     purrr_0.3.4
#  [6] readr_2.0.2     tidyr_1.1.4     tibble_3.1.5    tidyverse_1.3.1 ggplot2_3.3.5
# 
# loaded via a namespace (and not attached):
#  [1] shape_1.4.6         tidyselect_1.1.1    xfun_0.29           haven_2.4.3
#  [5] colorspace_2.0-2    vctrs_0.3.8         generics_0.1.1      htmltools_0.5.2
#  [9] yaml_2.2.1          utf8_1.2.2          rlang_0.4.12        pillar_1.6.4
# [13] glue_1.4.2          withr_2.4.3         DBI_1.1.1           dbplyr_2.1.1
# [17] modelr_0.1.8        readxl_1.3.1        lifecycle_1.0.1     munsell_0.5.0
# [21] gtable_0.3.0        cellranger_1.1.0    rvest_1.0.2         GlobalOptions_0.1.2
# [25] evaluate_0.14       knitr_1.37          tzdb_0.2.0          fastmap_1.1.0
# [29] fansi_0.5.0         broom_0.7.10        Rcpp_1.0.7          scales_1.1.1
# [33] backports_1.3.0     jsonlite_1.7.2      fs_1.5.0            hms_1.1.1
# [37] packrat_0.7.0       digest_0.6.29       stringi_1.7.6       grid_4.0.4
# [41] cli_3.1.0           tools_4.0.4         magrittr_2.0.1      wesanderson_0.3.6
# [45] crayon_1.4.2        pkgconfig_2.0.3     ellipsis_0.3.2      xml2_1.3.2
# [49] rsconnect_0.8.24    reprex_2.0.1        lubridate_1.8.0     rstudioapi_0.13
# [53] assertthat_0.2.1    rmarkdown_2.11      httr_1.4.2          R6_2.5.1
# [57] compiler_4.0.4</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
