<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jason Yeung" />

<meta name="date" content="2021-12-25" />

<title>Analyzing ViReMa Output in R</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ViReMa in R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Tutorial</a>
</li>
<li>
  <a href="https://jasonyeung.shinyapps.io/virema/">ViReMa with Shiny</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="mailto:&lt;jasonyeung012@gmail.com&gt;">
    <span class="fa fa-envelope fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://github.com/jayeung12/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Analyzing ViReMa Output in R</h1>
<h4 class="author">Jason Yeung</h4>
<h4 class="date">12/25/2021</h4>

</div>

<div id="TOC">
<ul>
<li><a href="#getting-started-with-virema-in-r"><strong>Getting Started with ViReMa in R</strong></a></li>
<li><a href="#setting-up"><strong>Setting Up</strong></a></li>
<li><a href="#plotting-recombination-events"><strong>Plotting Recombination Events</strong></a><ul>
<li><a href="#scatterplot"><strong>Scatterplot</strong></a></li>
<li><a href="#coverage-plot"><strong>Coverage Plot</strong></a></li>
<li><a href="#circos-plot"><strong>Circos Plot</strong></a></li>
<li><a href="#nucelotide-usage-plot"><strong>Nucelotide Usage Plot</strong></a></li>
</ul></li>
<li><a href="#segmented-viruses"><strong>Segmented Viruses</strong></a></li>
</ul>
</div>

<hr />
<div id="getting-started-with-virema-in-r" class="section level2">
<h2><strong>Getting Started with ViReMa in R</strong></h2>
<p>This tutorial will guide users through analyzing viral recombination data from <a href="https://bioconda.github.io/recipes/virema/README.html">ViReMa</a> in R. For more details on using R, see <a href="http://rmarkdown.rstudio.com" class="uri">http://rmarkdown.rstudio.com</a>.</p>
<p>If you are completely new to R, download the latest version of RStudio before following along.</p>
<p><br></p>
</div>
<div id="setting-up" class="section level2">
<h2><strong>Setting Up</strong></h2>
<p>The first step will be to install and load packages:</p>
<pre class="r"><code>install.packages(c(&quot;ggplot2&quot;,&quot;tidyverse&quot;,&quot;circlize&quot;))
library(ggplot2)
library(tidyverse)
library(circlize)</code></pre>
<p>This allows us to use popular functions we’ll need for data visualization.</p>
<p>Now, let’s <strong>load some example data</strong> I’ve provided:</p>
<pre class="r"><code>#We&#39;ve provided some sample data for the purposes of this tutorial
#Load provided example data

txt1 = &quot;https://raw.githubusercontent.com/jayeung12/jayeung12.github.io/main/files/SARS-CoV-2_1.txt&quot;
txt2 = &quot;https://raw.githubusercontent.com/jayeung12/jayeung12.github.io/main/files/SARS-CoV-2_2.txt&quot;  
txtdf1=read.table(txt1,skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)
txtdf2=read.table(txt2,skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)

data = rbind(txtdf1,txtdf2)</code></pre>
<p>Alternatively, load in your own data and follow along with the rest of the tutorial. Naturally, the plots will not be exactly the same as the ones we show.</p>
<pre class="r"><code>#Or you can use your own data
#Load own data from local directory

#This loads in a single file
#skip excludes the first line of the file
data = read.table(&quot;--replace with the file path--.bed&quot;,skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)

#To load in multiple files all in the same directory, we can use the following:

filenames &lt;- list.files(&quot;--replace with the directory--&quot;, pattern=&quot;*.bed&quot;, full.names=TRUE)
filelist = list()
for(i in 1:length(filenames)){
  filelist[[i]] =   read.table(filenames[i],skip=1, sep=&quot;\t&quot;,stringsAsFactors=FALSE, quote=&quot;&quot;)
}

#Combines all of the files, row by row
data = do.call(rbind, filelist)</code></pre>
<p><br> The data is in .BED6 format but the file extension may just say .BED or .TXT. If that’s the case, you will need to match the file extension in the above code to load in your own data.</p>
<p><strong>Included in your files need to have at least 6 columns</strong> - these correspond to .<br />
<img src="files/img.png" alt="optional caption text" /></p>
<p><br></p>
</div>
<div id="plotting-recombination-events" class="section level2">
<h2><strong>Plotting Recombination Events</strong></h2>
<div id="scatterplot" class="section level3">
<h3><strong>Scatterplot</strong></h3>
<p>Now that we have the data loaded in, let’s make our first plot. We will use ‘ggplot2’ to make a scatterplot with acceptor and donor sites on the axes.</p>
<p><img src="index_files/figure-html/pressure-1.png" width="672" /></p>
<p>Note that the <code>echo = FALSE</code> parameter was added to the code chunk to prevent printing of the R code that generated the plot.</p>
<p><br></p>
</div>
<div id="coverage-plot" class="section level3">
<h3><strong>Coverage Plot</strong></h3>
<p>This plot serves as a coverage plot. It takes each recombination event and adds the number of reads at nucleotide indexes between the donor and acceptor site. Depending on if deletion or duplication events are being plotted, this shows areas of the genome which are either frequently removed or included in recombination events. <br></p>
</div>
<div id="circos-plot" class="section level3">
<h3><strong>Circos Plot</strong></h3>
<p>This plot is a Circos plot made using the ‘circlize’ package. It allows for recombination events to be visualized alongside annotations or other quantitative data.</p>
<p><strong>There are multiple places to find annotations for your viral genome.</strong> NCBI has annotations associated with sequences. Below shows the process for a popular SARS-CoV-2 reference sequence. <img src="files/exportNCBIanno.gif" alt="Accessing the annotations at UCSC" /></p>
<p><br> The UCSC Genome browser website has a comprehensive collection of annotations for SARS-CoV-2. Access the “Data Integrator” at <a href="https://genome.ucsc.edu/cgi-bin/hgIntegrator" class="uri">https://genome.ucsc.edu/cgi-bin/hgIntegrator</a>. Below shows the process of exporting data:</p>
<p><img src="files/exportUCSCanno.gif" alt="Accessing the annotations at UCSC" /> <br><br> Then, copy the data into Notepad or another text editor and save it as a .BED or .TXT file.</p>
<p><br></p>
</div>
<div id="nucelotide-usage-plot" class="section level3">
<h3><strong>Nucelotide Usage Plot</strong></h3>
<p>This plot shows the frequency of nucleotides proximal and distal to recombination sites. In order to generate these plots, you had to select the option to produce this output when performing the analysis with ViReMA. The necessary columns look like this: <br></p>
</div>
</div>
<div id="segmented-viruses" class="section level2">
<h2><strong>Segmented Viruses</strong></h2>
<p>Plotting events in segmented viruses requires some changes to the code we used above. .BED files will contain multiple reference sequence IDs in the first column corresponding to the different segments of the viral genome.</p>
<p>We can also make the Circos plot with multiple reference sequences.</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
